### PR TITLE
fix: sdk wrappers not having device token registered because of application lifecycle

### PR DESCRIPTION
**Pull request by Shahroz16**
closes: https://github.com/customerio/issues/issues/9641

Complete each step to get your pull request merged in. [Learn more about the workflow this project uses](https://github.com/customerio/customerio-ios/blob/develop/docs/dev-notes/GIT-WORKFLOW.md). 
- [ ] [Assign members of your team](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to review the pull request. 
- [ ] Wait for pull request status checks to complete. If there are problems, fix them until you see that [all status checks are passing](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fsymfony.com%2Fdoc%2F4.3%2F_images%2Fdocs-pull-request-symfonycloud.png&f=1&nofb=1). 
- [ ] Wait until the pull request has been reviewed *and approved* by a teammate
- [ ] After code reviews are approved and you determine this PR is ready to merge, select *Squash and Merge* button on this screen, leave the title and description to the default values, then merge the PR.

### CODE REVIEW

**Review by levibostian — APPROVED** [review]
Assuming that [this comment I left](https://github.com/customerio/issues/issues/9641#issuecomment-1507222266) is correct, then I approve of this PR. 

That comment is important to make sure that FCM behavior is what I am expecting.

**Comment by levibostian** [comment]
Adding labels to not merge this PR until all CI status checks on this PR runs successfully. This is especially important for the CocoaPods FCM app compiling to run successful for us to test the `@available(iOSApplicationExtension, unavailable)` annotation use.

**Comment by levibostian** [comment]
Update: This PR is still blocked and should not be merged. Shahroz is changing the implementation as the previous solution did not work for FCM.

**Comment by Shahroz16** [comment]
@levibostian would appreciate another review here.

File: Sources/Common/DIGraph.swift
Comment (levibostian) [inline Sources/Common/DIGraph.swift:43]: I suggest we remove the optional casting code from this function and instead do a force cast, `as!`. 

This function is only called during automated tests and the force cast will only fail if the developer made a mistake setting up the dependency override. If tests crash, it's easy for the developer to catch the override issue and fix it. Otherwise, the override could go unnoticed and cause tests to be much harder to debug.

File: Sources/MessagingPush/MessagingPush.swift
Comment (levibostian) [inline Sources/MessagingPush/MessagingPush.swift:62]:
```suggestion
        // The SDK might not yet have been initialized. Save the device token so when the SDK does get initialized, the token can get registered to a profile. 
        globalDataStore.pushDeviceToken = deviceToken
        implementation?.registerDeviceToken(deviceToken)
```

Are there any consequences to simplifying the logic of this function to these 2 lines of code?

File: Sources/Tracking/ModuleTopLevelObject.swift
Comment (levibostian) [inline Sources/Tracking/ModuleTopLevelObject.swift:23]: Does `globalDataStore` need to be part of this `init` function and class? 

Since I don't see this class using this dependency, I would prefer we don't keep an instance of it in the class.

**Comment by levibostian** [comment]
Changed title of PR because I believe this is a bug fix and not a feature. 

Of course, if anyone believes this is not a bug fix, change the title to suggest something different.

File: Sources/Common/DIGraph.swift
Comment (Shahroz16) [reply Sources/Common/DIGraph.swift:43]: hmm, can do that, even though I would prefer logging an error than throwing it.  
We actually use the `getOverriddenInstance` in the sample app at times to override the logger etc. But open to change.

File: Sources/Tracking/ModuleTopLevelObject.swift
Comment (Shahroz16) [reply Sources/Tracking/ModuleTopLevelObject.swift:23]: `MessagingPush` is a singleton class. So to make it injectable, and testable I think this was the ideal way. But I might not be fully integrated with the iOS structure. So if there is any other way I can make it injectable in `MessagingPush` class?

File: Sources/Common/DIGraph.swift
Comment (levibostian) [reply Sources/Common/DIGraph.swift:43]: 🤔 I think it would be a good idea to add a comment to the `overrides` property saying that it's suggested to not use overrides in the SDK source code but only in tests (and sample apps if there is a need). 

If we as a team agree that overriding a dependency should only be used for tests (aka: non production code in the SDK), then doing things such as force casting in a function like this is OK. 

---

Back to your original comment, I agree that a log would be more useful then simply a crash without a log explaining the reason for the crash. 

I would be OK to log a message and then call `fatalError` if the cast is not successful. As long as we cause a crash, I am OK with the solution.

File: Sources/MessagingPush/MessagingPush.swift
Comment (levibostian) [reply Sources/MessagingPush/MessagingPush.swift:62]: The iOS SDK [already stores the device token to global data storage](https://github.com/customerio/customerio-ios/blob/f462325f9774dba01bbfcd1a96d2f41b461eb02e/Sources/Tracking/CustomerIOImplementation.swift#L226) every time that `registerDeviceToken` is called on the SDK. 

Therefore, after this PR gets merged with the changes made to `CustomerIO` class:
```swift
if let deviceToken = globalDAtaStore.deviceToken {
  registerDeviceToken(deviceToken)
}
```

we will be performing a redundant call even if `MessagingPush.registerDeviceToken` doesn't get called. Each time the SDK gets initialized after registering a device token the first time in the app's history, the device token will get registered again. So I am not sure the logic in this function as the PR currently has will make a difference. 

I think it's ideal to avoid redundant network calls. I would prefer it myself. I am not sure that we can accomplish that at this moment without adding a lot of extra logic to the SDK code?

File: Sources/Tracking/ModuleTopLevelObject.swift
Comment (levibostian) [reply Sources/Tracking/ModuleTopLevelObject.swift:23]: I believe if you modify [the constructor of `MessagingPush`](https://github.com/customerio/customerio-ios/blob/3dbc9bd9e68c8fab57d8c13578f3be1d89387c17/Sources/MessagingPush/MessagingPush.swift#L13-L14) to something like this:

```swift
public class MessagingPush: ModuleTopLevelObject<MessagingPushInstance>, MessagingPushInstance {
    @Atomic public private(set) static var shared = MessagingPush()
	private let globalDataStore: GlobalDataStore

    // testing constructor
    override internal init(implementation: MessagingPushInstance?, globalDataStore: GlobalDataStore, sdkInitializedUtil: SdkInitializedUtil) {
        super.init(implementation: implementation, sdkInitializedUtil: sdkInitializedUtil)
		self.globalDataStore = globalDataStore
    }
```

Then you wouldn't need to include `globalDataStore` in this `ModuleTopLevelObject` class.

Comment (Shahroz16) [reply Sources/Tracking/ModuleTopLevelObject.swift:23]: I don't think that will work, a subclass constructor cannot have more parameters than its superclass constructor.
It will return `Initializer does not override a designated initializer from its superclass`

File: Sources/MessagingPush/MessagingPush.swift
Comment (Shahroz16) [reply Sources/MessagingPush/MessagingPush.swift:62]: I am not sure I follow, the following line of code in `MessagingPush.registerDeviceToken` 
`if deviceToken != globalDataStore.pushDeviceToken`

It saves the extra network call. 

The later logic just saved another insertion to "GlobablDataStore" which will already happen in [the method you shared](https://github.com/customerio/customerio-ios/blob/f462325f9774dba01bbfcd1a96d2f41b461eb02e/Sources/Tracking/CustomerIOImplementation.swift#L226)

File: Sources/Common/DIGraph.swift
Comment (Shahroz16) [reply Sources/Common/DIGraph.swift:43]: Sure, I'll change it.

File: Sources/MessagingPush/MessagingPush.swift
Comment (levibostian) [reply Sources/MessagingPush/MessagingPush.swift:62]: Ok. Now that I think about it again, I can see how this would work. 

Thanks for making the comments in the function explaining why we are doing this logic.

File: Sources/Tracking/ModuleTopLevelObject.swift
Comment (levibostian) [reply Sources/Tracking/ModuleTopLevelObject.swift:23]: I am trying something out in Xcode right now. I'll update this comment or push a commit if I find something that will work.

Comment (levibostian) [reply Sources/Tracking/ModuleTopLevelObject.swift]: [Found a solution](https://github.com/customerio/customerio-ios/pull/285/commits/01f66a437a1596e10e45cce78e3a4d517dee0204#diff-d322168af6d12b45ff1a57021d6a686396fe280ff72e00f2c1872bc9e4abff3fR11-R21). 

As always, we can revert this commit if we don't like the solution. 

Otherwise, let's resolve this conversation.

**Review by levibostian — APPROVED** [review]

**Comment by Shahroz16** [comment]
@levibostian could use a re-review here after the merge. I don't like the fact we are using `sdkConfig` for the KeyValueStorage and internally just using `siteID`. But i didn't make the change here, because you specifically removed it so wanted to have your opinion first. Would do it in a separate PR.

**Review by levibostian — APPROVED** [review]

**Pull request merged**
